homeassistant/components/screenlogic/__init__.py

```diff
@@ -99,6 +85,44 @@ async def async_update_listener(hass: HomeAssistant, entry: ConfigEntry):
     await hass.config_entries.async_reload(entry.entry_id)
 
 
+def get_connect_info(hass: HomeAssistant, entry: ConfigEntry):
+    """Construct connect_info from configuration entry and returns it to caller."""
+    mac = entry.unique_id
+    # Attempt to re-discover named gateway to follow IP changes
+    discovered_gateways = hass.data[DOMAIN][DISCOVERED_GATEWAYS]
+    if mac in discovered_gateways:
+        connect_info = discovered_gateways[mac]
+    else:
+        _LOGGER.warning("Gateway rediscovery failed")
+        # Static connection defined or fallback from discovery
+        connect_info = {
+            SL_GATEWAY_NAME: name_for_mac(mac),
+            SL_GATEWAY_IP: entry.data[CONF_IP_ADDRESS],
+            SL_GATEWAY_PORT: entry.data[CONF_PORT],
+        }
+
+    return connect_info
+
+
+def get_new_gateway(hass: HomeAssistant, entry: ConfigEntry):
+    """Instantiate a new ScreenLogicGateway, connect to it and update."""
+
+    connect_info = get_connect_info(hass, entry)
+
+    try:
+        gateway = ScreenLogicGateway(**connect_info)
+    except ScreenLogicError as ex:
+        _LOGGER.error("Error while connecting to the gateway %s: %s", connect_info, ex)
+        raise ConfigEntryNotReady from ex
```

Comment:
I'm not sure we can consolidate into a single function. At least we need to handle the exceptions differently.

1. `ConfigEntryNotReady` may only be raised and leaked during config entry setup. Not afterwards.
2. `UpdateFailed` may only be raised and leaked within the update function of the coordinator.

If we still use a single function we need to catch the exceptions that occur out of place or time.

Reply:
I understand, i've refactored it into two separate functions.